packages/@react-stately/actiongroup/src/useActionGroupState.ts

```diff
@@ -11,23 +11,29 @@
  */
 
 import {ActionGroupCollection} from './';
-import {ActionGroupState, ActionGroupStateBase} from './types';
-import {MultipleSelection} from '@react-types/shared';
-import React, {useMemo} from 'react';
+import {ActionGroupProps} from '@react-types/actiongroup';
+import {ActionGroupState} from './types';
+import {CollectionBase, MultipleSelection} from '@react-types/shared';
+import {CollectionBuilder} from '@react-stately/collections';
+import {Key, useMemo} from 'react';
 import {SelectionManager, useMultipleSelectionState} from '@react-stately/selection';
 
-export function useActionGroupState(props: ActionGroupStateBase & MultipleSelection): ActionGroupState {
+export function useActionGroupState<T>(props: ActionGroupProps<T> & MultipleSelection & CollectionBase<T>): ActionGroupState<T> {
   let selectionState = useMultipleSelectionState(props);
 
+  let disabledKeys = useMemo(() =>
+    props.disabledKeys ? new Set(props.disabledKeys) : new Set<Key>()
+  , [props.disabledKeys]);
+
+  let builder = useMemo(() => new CollectionBuilder<T>(props.itemKey), [props.itemKey]);
   let collection = useMemo(() => {
-    let childrenArray = React.Children.toArray(props.children);
-    childrenArray = childrenArray.map(child => React.cloneElement(child, {
-      isSelected: selectionState.selectedKeys.has(child.key)
+    let nodes = builder.build(props, (key) => ({
+      isSelected: selectionState.selectedKeys.has(key),
+      isDisabled: disabledKeys.has(key) || props.isDisabled
```

Comment:
trying to figure out why provider didnt pass isDisabled to children (and it works with isEmphasized and isQuiet)

Reply:
Is the above not working? Haven't dig too much into it but it seems like the above would make each item `isDisabled` if its key exists in the `disabledKey` array provided to ActionGroup or if ActionGroup itself has `isDisabled` provided to it.  Or are you referring to a case where there is another Provider wrapping your ActionGroup?

Reply:
my bad, Provider works fine... it will pass isDIsabled, and it overrides context props with an item props https://github.com/adobe-private/react-spectrum-v3/blob/master/packages/%40react-spectrum/provider/src/Provider.tsx#L183
`Object.assign({}, {context}, {itemProps})`

I just thought it will work without` || props.isDisabled ` ... and that context props have priority over itemProps